### PR TITLE
[Federation] Add an end-to-end test verifying that deleting a federated namespace deletes child replicasets.

### PR DESCRIPTION
**Pull request by perotinus**
Verifies #38225.

Also, remove a few custom package aliases.

### CODE REVIEW

**Comment by perotinus** [comment]
cc @madhusudancs

**Comment by perotinus** [comment]
I signed it!

**Comment by madhusudancs** [comment]
A couple of minor nits but mostly looks good.

Also, please rebase.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[test/e2e_federation/federated-namespace.go, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-Kcf3ev75nYfHaAlWR_o:-Kcf3ev75nYfHaAlWR_p:b-zyc7k) ([raw file](https://github.com/kubernetes/kubernetes/blob/1bd140812bf8813f5fe5d33b7095374253f6d952/test/e2e_federation/federated-namespace.go#L103)):*
> ```Go
> 
> 			nsName := createNamespace(f.FederationClientset.Core().Namespaces())
> 			rsName := v1.SimpleNameGenerator.GenerateName(eventNamePrefix)
> ```

By convention, `eventNamePrefix` is used for Event API type object names. Please define another const for replicasets and use that const here.

---

*[test/e2e_federation/federated-namespace.go, line 143 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-KcfLg623K9-TWHM13_f:-KcfLg63gcKfYqwecYpw:b-sp1b1) ([raw file](https://github.com/kubernetes/kubernetes/blob/1bd140812bf8813f5fe5d33b7095374253f6d952/test/e2e_federation/federated-namespace.go#L143)):*
> ```Go
> 
> 			By(fmt.Sprintf("Verify that replicaset %s was deleted as well", rsName))
> 			err = wait.Poll(5*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
> ```

I recommend pulling this and https://github.com/kubernetes/kubernetes/blob/a50ecde91951ac79774a17d8de49dad4d4ea1c99/test/e2e_federation/federated-replicaset.go#L417 into a common helper function.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-:-KcfNjX3jJ3voeO__DaM:b-qf5mrz)*

**Comment by perotinus** [comment]
Thanks, @madhusudancs! PTAL when you get a chance.

**Comment by perotinus** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e_federation/federated-namespace.go, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-Kcf3ev75nYfHaAlWR_o:-KctY6wqmTUh98zvX_tv:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/1bd140812bf8813f5fe5d33b7095374253f6d952/test/e2e_federation/federated-namespace.go#L103)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

By convention, `eventNamePrefix` is used for Event API type object names. Please define another const for replicasets and use that const here.
</blockquote></details>

Done.

---

*[test/e2e_federation/federated-namespace.go, line 143 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-KcfLg623K9-TWHM13_f:-KctY7RfD7wpHPJ8mSKb:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/1bd140812bf8813f5fe5d33b7095374253f6d952/test/e2e_federation/federated-namespace.go#L143)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I recommend pulling this and https://github.com/kubernetes/kubernetes/blob/a50ecde91951ac79774a17d8de49dad4d4ea1c99/test/e2e_federation/federated-replicaset.go#L417 into a common helper function.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278)*

**Comment by madhusudancs** [comment]
Reviewed 2 of 3 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e_federation/federation-util.go, line 530 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-KctaYxYcfSWf-DSq2Py:-KctaYxZ5ZOjBs8plA1l:b-5dxpmd) ([raw file](https://github.com/kubernetes/kubernetes/blob/089c58938979aa797ce2c9b56cec45ae3e4f3e9a/test/e2e_federation/federation-util.go#L530)):*
> ```Go
> }
> 
> /*
> ```

It is a convention to use // on each line for docstrings.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278)*

**Comment by perotinus** [comment]
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e_federation/federation-util.go, line 530 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-KctaYxYcfSWf-DSq2Py:-KctaqUAvsxH7y98U5Jr:b-mdrf0p) ([raw file](https://github.com/kubernetes/kubernetes/blob/089c58938979aa797ce2c9b56cec45ae3e4f3e9a/test/e2e_federation/federation-util.go#L530)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

It is a convention to use // on each line for docstrings.
</blockquote></details>

Ah, OK. Done.

Should we fix the other comments in this file, perhaps in a follow-up PR?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278)*

**Comment by madhusudancs** [comment]
Reviewed 1 of 3 files at r2, 1 of 2 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/e2e_federation/federation-util.go, line 530 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-KctaYxYcfSWf-DSq2Py:-KctbwMPpIGS7kmuZbSI:bpm2kac) ([raw file](https://github.com/kubernetes/kubernetes/blob/089c58938979aa797ce2c9b56cec45ae3e4f3e9a/test/e2e_federation/federation-util.go#L530)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Ah, OK. Done.

Should we fix the other comments in this file, perhaps in a follow-up PR?
</blockquote></details>

Follow up PR for other instances would be great. I will LGTM and approve this PR.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278)*

**Comment by madhusudancs** [comment]
/lgtm
/approve

---

Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-:-KctcEHc5ChJ_Sng8vDS:b6yv7yi)*

**Comment by perotinus** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/e2e_federation/federation-util.go, line 530 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278#-KctaYxYcfSWf-DSq2Py:-KcteaAZePAe-zCOWIkt:b-hpdfgu) ([raw file](https://github.com/kubernetes/kubernetes/blob/089c58938979aa797ce2c9b56cec45ae3e4f3e9a/test/e2e_federation/federation-util.go#L530)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Follow up PR for other instances would be great. I will LGTM and approve this PR.
</blockquote></details>

Thanks! I'll start working on that now.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41278)*

**Comment by perotinus** [comment]
@k8s-bot verify test this

**Comment by madhusudancs** [comment]
@k8s-bot ok to test

**Comment by perotinus** [comment]
@madhusudancs Can you re-LGTM?

File: test/e2e_federation/federated-namespace.go [hunk 1]
```diff
@@ -97,18 +98,67 @@ var _ = framework.KubeDescribe("Federation namespace [Feature:Federation]", func
...
+			}
+
+			By(fmt.Sprintf("Deleting namespace %s", nsName))
+			deleteNamespace(nil, nsName,
+				f.FederationClientset.Core().Namespaces().Get,
+				f.FederationClientset.Core().Namespaces().Delete)
+
+			By(fmt.Sprintf("Verify that replicaset %s was deleted as well", rsName))
+
+			waitForReplicaSetToBeDeletedOrFail(f.FederationClientset, nsName, rsName)
```
Comment (nikhiljindal) [inline test/e2e_federation/federated-namespace.go]: We should not require a wait here. Replicaset should have been deleted before the namespace was deleted.

Comment (madhusudancs) [inline test/e2e_federation/federated-namespace.go]: You mean the delete namespace operation is synchronous, i.e. it won't return until the namespace cleanup is complete?

Comment (nikhiljindal) [inline test/e2e_federation/federated-namespace.go]: the function that this test is using to delete namespace is synchronous. It waits for the namespace to be deleted before returning.
So here, we can directly check that the replicaset is gone. No need to wait again.

Comment (csbell) [inline test/e2e_federation/federated-namespace.go]: After following this commentary, I'd suggest appending a suffix to deleteNamespace to indicate that it is synchronous (+Sync or +AndWait, whichever....).

Comment (madhusudancs) [inline test/e2e_federation/federated-namespace.go]: @nikhiljindal makes sense.

@csbell good idea.

@perotinus is this something that you can fix?

Comment (madhusudancs) [inline test/e2e_federation/federated-namespace.go]: We use `AndWait` in kubernetes tests often, so we can stick with that I guess.

**Pull request merged**